File: frontend/src/shared/lib/display-format.ts
/** Formato de presentación CL (montos, RUT, folios). */

function isBlank(value: unknown): boolean {
  if (value == null) return true;
  const s = String(value).trim();
  return !s || ["nan", "none", "nat", "n/a"].includes(s.toLowerCase());
}

export function formatFolio(value: unknown): string {
  if (isBlank(value)) return "";
  const n = Number(String(value).replace(",", "."));
  if (Number.isFinite(n)) return String(Math.trunc(n));
  return String(value).trim();
}

export function formatMontoCl(value: unknown): string {
  if (isBlank(value)) return "";
  let s = String(value).trim().replace(/\$/g, "").replace(/\s/g, "").replace(/\.-$/, "");
  if (s.endsWith(".") && s.indexOf(".") === s.length - 1) s = s.slice(0, -1);
  if (s.includes(",") && s.includes(".")) {
    s = s.lastIndexOf(",") > s.lastIndexOf(".") ? s.replace(/\./g, "").replace(",", ".") : s.replace(/,/g, "");
  } else if (s.includes(",")) {
    const parts = s.split(",");
    if (parts.length > 1 && parts.every((p) => /^\d+$/.test(p)) && parts.slice(1).every((p) => p.length === 3)) {
      s = parts.join("");
    } else {
      s = s.replace(",", ".");
    }
  } else if (s.includes(".")) {
    const parts = s.split(".");
    if (parts.length > 1 && parts.every((p) => /^\d+$/.test(p)) && parts.slice(1).every((p) => p.length === 3)) {
      s = parts.join("");
    }
  }
  const n = Number(s);
  if (!Number.isFinite(n)) return String(value).trim();
  const body = Math.trunc(n).toLocaleString("es-CL");
  return `$${body}.-`;
}

export function formatRutCl(value: unknown): string {
  if (isBlank(value)) return "";
  let raw = String(value).trim();
  if (/^[\d.]+-[\dkK]$/.test(raw) && (raw.match(/\./g)?.length ?? 0) >= 1) return raw;
  if (/^\d+\.0+$/.test(raw)) raw = raw.split(".", 1)[0];
  else if (typeof value === "number" && Number.isInteger(value)) raw = String(value);
  const normalized = raw.replace(/[.\-\s\u00A0]/g, "").toUpperCase();
  if (normalized.length < 2) return raw.replace(/\.0$/, "");
  const dv = normalized.slice(-1);
  const cuerpo = normalized.slice(0, -1);
  const rev = cuerpo.split("").reverse().join("");
  const chunks: string[] = [];
  for (let i = 0; i < rev.length; i += 3) chunks.push(rev.slice(i, i + 3));
  const cuerpoFmt = chunks.map((c) => c.split("").reverse().join("")).reverse().join(".");
  return `${cuerpoFmt}-${dv}`;
}

export function formatDisplayField(value: unknown, kind: "folio" | "monto" | "rut" | "text" = "text"): string {
  if (isBlank(value)) return "";
  switch (kind) {
    case "folio":
      return formatFolio(value);
    case "monto":
      return formatMontoCl(value);
    case "rut":
      return formatRutCl(value);
    default:
      return String(value).trim();
  }
}

File: lib/display_format.py
"""Formato de presentación para UI (montos CLP, RUT, folios)."""
from __future__ import annotations

import re
from typing import Any

import utils


def _is_blank(value: Any) -> bool:
    if value is None:
        return True
    try:
        import pandas as pd

        if pd.isna(value):
            return True
    except Exception:
        pass
    s = str(value).strip()
    return not s or s.lower() in {"nan", "none", "nat", "n/a"}


def format_folio(value: Any) -> str:
    """Número de boleta / folio sin decimales."""
    if _is_blank(value):
        return ""
    try:
        return str(int(float(str(value).replace(",", "."))))
    except (TypeError, ValueError):
        return str(value).strip()


def format_monto_cl(value: Any, *, suffix: bool = True) -> str:
    """Monto chileno: $101.916.-"""
    if _is_blank(value):
        return ""
    try:
        s = str(value).strip().replace("$", "").replace(" ", "").replace(".-", "")
        if s.endswith(".") and s.count(".") == 1:
            s = s[:-1]
        if "," in s and "." in s:
            if s.rfind(",") > s.rfind("."):
                s = s.replace(".", "").replace(",", ".")
            else:
                s = s.replace(",", "")
        elif "," in s:
            parts = s.split(",")
            if (
                len(parts) > 1
                and all(p.isdigit() for p in parts)
                and all(len(p) == 3 for p in parts[1:])
            ):
                s = "".join(parts)
            else:
                s = s.replace(",", ".")
        elif "." in s:
            parts = s.split(".")
            if (
                len(parts) > 1
                and all(p.isdigit() for p in parts)
                and all(len(p) == 3 for p in parts[1:])
            ):
                s = "".join(parts)
        n = int(float(s))
        body = f"{n:,}".replace(",", ".")
        return f"${body}.-" if suffix else f"${body}"
    except (TypeError, ValueError):
        return str(value).strip()


def format_rut_cl(value: Any) -> str:
    """RUT con puntos y guión: 65.175.239-6"""
    if _is_blank(value):
        return ""
    raw = str(value).strip()
    if re.match(r"^[\d\.]+-[\dkK]$", raw) and raw.count(".") >= 1:
        return raw
    # Excel suele entregar RUT+DV como float (651752396.0 → 65175239-6).
    if re.match(r"^\d+\.0+$", raw):
        raw = raw.split(".", 1)[0]
    elif isinstance(value, float):
        try:
            if float(value).is_integer():
                raw = str(int(value))
        except (TypeError, ValueError):
            pass
    normalized = utils.normalizar_rut_con_dv(raw)
    if not normalized or len(normalized) < 2:
        return raw.replace(".0", "").strip()
    cuerpo, dv = normalized[:-1], normalized[-1]
    rev = cuerpo[::-1]
    chunks = [rev[i : i + 3] for i in range(0, len(rev), 3)]
    cuerpo_fmt = ".".join(chunk[::-1] for chunk in chunks[::-1])
    return f"{cuerpo_fmt}-{dv}"


def format_rut_sin_dv(value: Any) -> str:
    if _is_blank(value):
        return ""
    digits = utils.normalizar_rut_digits(str(value))
    if not digits:
        return str(value).strip().replace(".0", "")
    try:
        return str(int(digits))
    except ValueError:
        return digits
